### PR TITLE
[JENKINS-57761] - Add abstract test for export/configure roundtrip

### PR DESCRIPTION
**Pull request by MRamonLeon**
### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!

- [x] Please describe what you did

- [x] Link to relevant GitHub issues or pull requests
https://github.com/jenkinsci/metrics-plugin/pull/45 All code in the metrics plugin
https://github.com/jenkinsci/metrics-plugin/pull/46 The implementation in the metrics plugin

- [x] Link to relevant [Jenkins JIRA issues](https://issues.jenkins-ci.org)
[JENKINS-57761](https://issues.jenkins-ci.org/browse/JENKINS-57761) Developed while creating the roundtrip test for metrics plugin.

- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### Description

Add a new abstract test to be used by plugins as a full export / configure test. The goal is to make as easy as possible the life to plugins authors and at the same time, have tested as many plugins as possible. With this PR, plugin authors only need to extend a class and override 3 methods. See https://github.com/jenkinsci/metrics-plugin/pull/46

Currently, there is a problem exporting the full configuration and importing it. The _**defaultProperties**_ are exported and make the file exported to be not importable. See https://issues.jenkins-ci.org/browse/JENKINS-57122 I've done a workaround by using the config piece of the plugin under test, but it should be removed once the problem is solved.

### CODE REVIEW

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    public RestartableJenkinsRule r = new RestartableJenkinsRule();
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+
+    @Rule
+    public LoggerRule logging = new LoggerRule();
+
+    static {
+        // The restartable rule add spaces to Jenkins home and it makes the second and next steps to fail loading the
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: is this something that should be fixed?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        assertNotNull(resourceContent);
+
+        r.then(step -> {
+            // Configure and validate
+            configureWithResource(resourcePath);
+            assertConfiguredAsExpected(r, resourceContent);
+
+            // Get the full configuration
+            //String jenkinsConf = getJenkinsConfViaWebUI();
+            //hack: the full Jenkins config fails due to defaultProperties of maven, we use the config of the plugin
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: there's a PR for it but seems unlikely to be merged ☹️
may need to hack something into jcasc to blacklist it
https://github.com/jenkinsci/jenkins/pull/3992

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ *
+ */
+public abstract class ExportImportRoundTripAbstractTest {
```
Comment (casz) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Could you explain the reasoning for a restartable Jenkins with roundtrip configuration? 
What's the issue? What are you trying to solve?
Why can't you use the existing test method that others have done?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 1 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Fixed

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 2 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I've added a workaround, but would it be great if it works. We then can do the full export / replace roundtrip :smiley:

File: integrations/src/test/java/io/jenkins/plugins/casc/GlobalMatrixAuthorizationTest.java [hunk 4]
```diff
@@ -1,35 +0,0 @@
...
-import org.junit.Rule;
-import org.junit.Test;
-
-import static org.junit.Assert.assertEquals;
-
-/**
- * @author Mads Nielsen
- * @since TODO
- */
-public class GlobalMatrixAuthorizationTest {
```
[the remark this answers is not in the document]
Comment (MRamonLeon) [reply integrations/src/test/java/io/jenkins/plugins/casc/GlobalMatrixAuthorizationTest.java]: I deleted these two files by mistake, sorry for the confusion

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 3 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I'm trying to test easily for a plugin author:
1. The instance is configured with his/her piece of yaml
2. The full configuration exported can be applied (not currently - added a hack)
3. The instance is configured using the default config file.

**Comment by MRamonLeon** [comment]
@oleg-nenashev @batmat @alecharp @fcojfernandez @varyvol A useful abstract test to go faster when checking plugins compatibility with JCasC.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+        r.then(step -> {
+            try {
+                // Verify the log shows it's configured
+                assertLogAsExpected(stringInLogExpected());
+
+                // Verify the configuration set at home/jenkins.yaml is loaded
+                assertConfiguredAsExpected(r, resourceContent);
+            } finally {
+                restoreJenkinsConfig();
```
Comment (varyvol) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: What if the failure happens during the first step?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+     * 6.  Check the Jenkins configuration is valid via Web UI (used the plugin config so far).
+     * 7.  Apply the Jenkins configuration via Web UI (maybe not needed, but we test it as well).
+     * 8.  Write the Jenkins configuration to $JENKINS_ROOT/jenkins.yaml.
+     * 9.  Restart Jenkins.
+     * 10. Check the {@link #stringInLogExpected()} is set during the restart.
+     *
+     * @throws IOException If an exception is thrown managing resources or files.
+     */
+    @Test
+    public void exportImportRoundTrip() throws IOException {
```
Comment (fcojfernandez) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: nit: I've seen similar tests named as `configRoundtrip` in the past. Doesn't it suit here?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    public abstract String configResource();
+
+    /**
+     * Return the string that should be in the logs of the JCasC logger to verify it's configured after a restart. This
+     * string should be unique to avoid interpreting that it was configured successfully, but it wasn't.
+     * @return the unique string to be in the logs to certify the configuration was done successfully.
+     */
+    public abstract String stringInLogExpected();
+
+    @Before
```
Comment (fcojfernandez) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I think it can be dangerous. If someone extends this class and adds other tests that are not related with the subject of this abstract class, this method would be executed always. Maybe it's worth to call it from the test method

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+        cascJenkinsConfigPropOld = System.getProperty(ConfigurationAsCode.CASC_JENKINS_CONFIG_PROPERTY, null);
+    }
+
+    /*
+    Annotated with @After is executed before the second step is ended.
+    We need to call it at the end of the second step.
+     */
+    //@After
```
Comment (fcojfernandez) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: If it's not used, better remove it?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 3 above)
Comment (fcojfernandez) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I think the idea is great. I would change maybe some points of the implementation since I've said in the general comment, but that should help to any maintainer to test the compatibility of their plugins.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 5 above)
Comment (fcojfernandez) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: My understanding is that each execution should have its own Jenkins Rule running, so it should not impact. The test only would fail. Am I right @MRamonLeon ?

Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: You're right. But the `System.setProperty` may affect other tests. Anyway, I've removed the setter of the property after other fixes because it was failing from the beginning and it seems to work fine. So, after removing that I can safely remove the backup and restore code.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+     * supposedly loaded are passed.
+     */
+    protected abstract void assertConfiguredAsExpected(RestartableJenkinsRule j, String configContent);
+
+    /**
+     * Return the resource path (yaml file) to be loaded. i.e: If the resource is in the same package of the implementor
+     * class, then: my-config.yaml
+     * @return the resource name and path.
+     */
+    protected abstract String configResource();
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: can't this be defaulted?
```suggestion
    protected String configResource() {
        return "configuration-as-code.yaml";
    }
```

Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Could be. It would be great to add the requirement to the documentation as well.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+ * 4.  Export the schema via Web UI (commented out as there is no good validator so far).
+ * 5.  Verify the Jenkins configuration against the schema (commented out as there is no good validator so far).
+ * 6.  Check the Jenkins configuration is valid via Web UI (used the plugin config so far).
+ * 7.  Apply the Jenkins configuration via Web UI (maybe not needed, but we test it as well).
+ * 8.  Write the Jenkins configuration to $JENKINS_ROOT/jenkins.yaml.
+ * 9.  Restart Jenkins.
+ * 10. Check the {@link #stringInLogExpected()} is set during the restart.
+ *
+ * All the plugin author needs to do is override the methods providing:
+ * 1. The resource with the yaml configuration of the plugin
```
Comment (fcojfernandez) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Now it's optional and it's not mandatory to override
```suggestion
 * 1. The resource with the yaml configuration of the plugin in case they use their own name for the file
```

File: pom.xml [hunk 11]
```diff
@@ -93,6 +93,16 @@
           </execution>
         </executions>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <excludes>
+            <exclude>**/ProjectMatrixAuthorizationTest.java</exclude>
```
Comment (casz) [inline pom.xml]: WHY?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 2 above)
Comment (timja) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Without that working I'm not that comfortable with this PR, i.e if a plugin user implements this test then there's no coverage for the exporting of their plugin.

The the bit which seems to be missing is without asserting the export is what you expect you risk the plugin author:
1. not viewing the export and realising it's not what they expect
2. unintentionally breaking the config in an update but not realising because the round trip works

File: pom.xml (hunk 11 above)
Comment (MRamonLeon) [reply pom.xml]: Again, I pushed it by mistake, sorry. For some reason, these two tests are not running on my laptop. That's why I'm so obsessed with removing them.

**Comment by MRamonLeon** [comment]
We have 3 plugins using this feature so far:
* https://github.com/jenkinsci/metrics-plugin/pull/46
* https://github.com/jenkinsci/ssh-credentials-plugin/pull/42
* https://github.com/jenkinsci/ssh-slaves-plugin/pull/126

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 2 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Hi Tim. I've tried to check the exported configuration has the same values as the YAML used. But, again, I faced a bug and I've commented out the code.

When running Jenkins normally:
You can add secrets ciphered in the YAML. CasC uploads it without reciphering. And when you export the configuration, the data is the same as it was in the yaml file.

When running with mvn running tests or hpi:run:
CasC ciphers the secrets, no matter if they were already ciphered in the yaml.

This bug makes impossible for plugins with secrets in the yaml to pass the test. We can solve this issue to uncomment this hack, or fix the full export to uncomment both hacks I did. :-)

Anyway, I've pushed the code.

Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I've added the RoundTripMailerTest to prove it's working fine (when no secrets in the yaml)

**Comment by MRamonLeon** [comment]
@casz is there anything else I can do to get it merged? Although this test has hacks it had helped a lot to find some bugs. Maybe it's good to merge it but avoid adding it to the documentation until the bugs are solved.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+            //but first, we guarantee that the exported configuration is the same as the configured.
+            //TODO: remove when https://issues.jenkins-ci.org/browse/JENKINS-57122 is solved
+            //@Issue("JENKINS-57122")
+            jenkinsConf = getResourceContent(resourcePath);
+
+            // Get the schema
+            //String schema = getSchemaViaWebUI();
+
+            // Verify it's compliant to the schema
+            // TODO: when YAML schema validation is mature enough we can do that.
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: I think we can remove all the commented out code, make's it more difficult to see what's actually being done in the present.
YAML schema is a long way off working its current state

Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: ok

**Comment by timja** [comment]
If you're happy with this please remove WIP

**Comment by MRamonLeon** [comment]
The test using the abstract class and checking the result of the export against the YAML used is green:
Screenshot from 2019-06-06 11-30-37

Now, I'm going to comment the check out to avoid tests with secrets to fail.

**Comment by MRamonLeon** [comment]
It would be great if we can merge and release it. Is there anything else I can do to achieve that?

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+ * 6.  Check the Jenkins configuration is valid via Web UI (used the YAML config because of the bug JENKINS-57122).
+ * 7.  Apply the Jenkins configuration via Web UI (maybe not needed, but we test it as well).
+ * 8.  Write the Jenkins configuration to $JENKINS_ROOT/jenkins.yaml.
+ * 9.  Restart Jenkins.
+ * 10. Check the {@link #stringInLogExpected()} is set during the restart.
+ *
+ * All the plugin author needs to do is override the methods providing:
+ * 1. The resource with the yaml configuration of the plugin in case they use their own name for the file.
+ * 2. A way to validate the configuration is established.
+ * 3. A string that should be present in the logs that guarantees the config is loaded. Usually a weird text configured.
```
Comment (oleg-nenashev) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]:
```suggestion
 * 3. A string that should be present in the logs that guarantees the config is loaded. Usually a weird text configured.
 *
 * @since TODO
```

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+        r.then(step -> {
+            // Verify the log shows it's configured
+            assertLogAsExpected(stringInLogExpected());
+
+            // Verify the configuration set at home/jenkins.yaml is loaded
+            assertConfiguredAsExpected(r, resourceContent);
+        });
+    }
+
+//    private void assertJenkinsHasConfiguredPath(String jenkinsConfig, String resourcePath) throws IOException, URISyntaxException {
```
Comment (oleg-nenashev) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: If you comment out a big chunk of code, please put explanation, or a link to the discussion in comments before it.
Bonus points for TODO

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+        WebRequest request = new WebRequest(client.createCrumbedUrl("configuration-as-code/checkNewSource"), POST);
+        NameValuePair param = new NameValuePair("newSource", f.toURI().toURL().toExternalForm());
+        request.setRequestParameters(Collections.singletonList(param));
+        WebResponse response = client.loadWebResponse(request);
+        assertEquals(200, response.getStatusCode());
+        String res = response.getContentAsString();
+        assertThat(res, containsString("The configuration can be applied"));
+    }
+
+    private void applyConfigViaWebUI(String jenkinsConfig) throws Exception {
```
Comment (oleg-nenashev) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: Maybe later we could have JCasC methods, e,g, in https://github.com/jenkinsci/java-client-api

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 14 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: done

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java (hunk 15 above)
Comment (MRamonLeon) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java]: it would be great

**Comment by MRamonLeon** [comment]
@timja, @varyvol I don't want to lose these pieces of code. We only need to uncomment that after fixing bugs. 

If merging without squashing, I'll be happy to remove all the commented out code.

**Comment by timja** [comment]
Your PR will always be here and we can resurrect the code from this, we can do a rebase and merge though

**Comment by casz** [comment]
Squash and merge does not squash the commits inside the PR, and your branch and PR will still have the code and history as @timja said.
So I'd prefer if you could remove the commented code 😓

**Comment by casz** [comment]
@MRamonLeon are we good to merge? 👍

**Comment by MRamonLeon** [comment]
I've removed the commented out code, although now the name of the class and method is not accurate because it's not doing an export/import. I've changed that as well because for now, I can change all the existing implementations.

**Comment by MRamonLeon** [comment]
uuuooohhhh you were hidden under the curtains :laughing:

**Comment by MRamonLeon** [comment]
Absolutely

**Comment by casz** [comment]
We are too fast for you 😅 I suggest using [refined-github](https://github.com/sindresorhus/refined-github) as it will do the magic and pull new in comments.

**Comment by casz** [comment]
Waiting on CI 😓

**Comment by MRamonLeon** [comment]
aham, so there was a trick, it wasn't just...
image

**Comment by timja** [comment]
image
@casz

**Comment by casz** [comment]
I know @timja not so useful when sitting in Train on mobile data... 😓 
Kinda want to remove LGTM it is simply too slow...

**Comment by MRamonLeon** [comment]
Ready!

**Comment by MRamonLeon** [comment]
Can you cut off a release, to avoid incremental in the plugins extending the test? Thank you.

**Comment by oleg-nenashev** [comment]
@casz if you are fine, I could use it as an opportunity to test the release flow on my machine

**Comment by casz** [comment]
@oleg-nenashev I think we are ready, I would ask you setup GPG signing of tags at the very least but going through with a release. 😓

**Comment by casz** [comment]
@MRamonLeon @fcojfernandez perhaps the missing things could be fixed now?
